lib/types.d.ts

```diff
@@ -0,0 +1,32 @@
+type MergeBy<TObject, TOtherObject> = Omit<TObject, keyof TOtherObject> & TOtherObject;
+
+type DeepRecord<TKey extends string | number | symbol, TValue> = {[key: string]: TValue | DeepRecord<TKey, TValue>};
+
+type DeepKeyOf<TObject> = TObject extends object
+    ? {
+          [TKey in keyof TObject & (string | number)]: TObject[TKey] extends Record<string, unknown> ? `${TKey}.${DeepKeyOf<TObject[TKey]>}` | TKey : TKey;
+      }[keyof TObject & (string | number)]
+    : TObject;
+
+type TypeOptions = MergeBy<
+    {
+        keys: string;
+        collectionKeys: string;
+        values: Record<string, unknown>;
+    },
+    CustomTypeOptions
+>;
+
+interface CustomTypeOptions {}
+
+type Key = TypeOptions['keys'];
+type CollectionKey = TypeOptions['collectionKeys'];
+type Value = TypeOptions['values'];
```

Comment:
@fabioh8010 What happens if we override values in the App but with some onyx keys missing? Would we get an error? Would all functions work as expected? 

example:
```ts
type Key = 'test1' | 'test2';
type Value = {
    test2: string
}
```

Reply:
That would be awesome to throw an error in case some keys are missing, without it contributors might forget to pass it and that could lead to bugs.

Reply:
@blazejkustra In this case `test1` will be inferred as `unknown` by Onyx methods and HOC, it won't break as long as the developer handle the type correctly.

> That would be awesome to throw an error in case some keys are missing, without it contributors might forget to pass it and that could lead to bugs.

I need to think about this, no ideas at the moment 😅

Reply:
I thought using `interface` will work but it didn't 😭 

```ts
type OnyxValuesSkeleton = { [K in OnyxKey | OnyxCollectionKey]: unknown };

interface OnyxValues extends OnyxValuesSkeleton {
  [ONYXKEYS.ACCOUNT]: Account;

  // Use this pattern to type collection keys.
  [download: `${typeof ONYXKEYS.COLLECTION.DOWNLOAD}${string}`]: Download;
  [report: `${typeof ONYXKEYS.COLLECTION.REPORT}${string}`]: Report;

  // [ONYXKEYS.IS_LOADING_PAYMENT_METHODS]: boolean; <- doesn't error even if you forget to provide value types for some keys 😢 
}
```

Reply:
@hayata-suenaga @blazejkustra Still didn't find a way to ensure such thing during type definition, but I think it's not so much of a problem as the developer will notice that `Onyx` methods are returning `unknown` for the key they forgot to add in `OnyxValues`. We can also add a comment next to `OnyxKey` and `OnyxValues` explaining about this. WDYT?

Reply:
that's fine with me, too 👍 thank you for tying to find a solution on this Fabio 🙇